qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -210,11 +210,12 @@ def run(self, dag):
 
         dag_list = []
         for node in dag.topological_op_nodes():
-            cargs = []
-            if node.op.condition is not None and not node.cargs:
-                cargs = [self._clbit_indices[x] for x in dag._bits_in_condition(node.op.condition)]
-            else:
-                cargs = [self._clbit_indices[x] for x in node.cargs]
+            cargs = [self._clbit_indices[x] for x in node.cargs]
+            if node.op.condition is not None:
+                for clbit in dag._bits_in_condition(node.op.condition):
+                    clbit_index = self._clbit_indices[clbit]
+                    if clbit_index not in cargs:
+                        cargs.append(clbit_index)
```

Comment:
How concerned about the quadratic runtime of this do you want to be?

Reply:
Heh, I was concerned about that which is why I wrote it the way I did in the original flawed implementation. In practice I don't think this will matter much because it depends on the number of cargs and clbits. To have this be really expensive we need an instruction with a lot of condition bits and a lot of classical bits. I think in practice we're likely to have a lot of either or but not both. But we do need to avoid duplicates because that would cause a parallel edge in the graph and throw off the output. That being said I can probably change `cargs` to be a set and I don't **think** that would break anything in the internal workings of sabre since we only use this as part of the edge list. Let me give that a try and see what happens.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/8727/commits/95db19c0a55cc8ca65213d9218c573fd9f66069b